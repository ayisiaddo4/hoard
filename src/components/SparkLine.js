import React from 'react';
import PropTypes from 'prop-types';
import { View, StyleSheet, Dimensions } from 'react-native';
import Svg, { Polyline } from 'react-native-svg';
import { getColors } from 'styles';

export default class SparkLine extends React.Component {
  static propTypes = {
    positive: PropTypes.bool.isRequired,
    children: PropTypes.arrayOf(PropTypes.number).isRequired,
  };

  state = {
    width: 0,
    height: 0,
  };

  handleLayout = ({ nativeEvent: { layout: { width, height } } }) => {
    this.setState({
      width: width ? width : Dimensions.get('window').width,
      height,
    });
  };

  render() {
    const { width, height } = this.state;
    const { positive, children } = this.props;

    const themeColors = getColors();

    const maxValue = Math.max(...children);
    const minValue = Math.min(...children);
    const range = maxValue - minValue;
    const distance = width / (children.length - 1);
    const renderablePoints = children
      .map((v, i) => ({
        y: (v - minValue) * height / range,
        x: distance * i + (i ? distance : 0),
      }))
      .map(v => `${v.x},${-v.y + height}`);

    return (
      <View onLayout={this.handleLayout} style={styles.container}>
        {!width || !height ? null : (
          <Svg height={`${height}`} width={`${width}`}>
            <Polyline
              points={renderablePoints.join(' ')}
              fill="none"
              stroke={
                positive ? themeColors.chartPositive : themeColors.chartNegative
              }
              strokeWidth="1"
            />
          </Svg>
        )}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'flex-end',
  },
  dot: {
    marginRight: 2,
    marginLeft: 2,
  },
});
