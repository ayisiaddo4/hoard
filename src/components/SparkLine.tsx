import * as React from 'react';
import {View, StyleSheet, Text, Dimensions, LayoutChangeEvent} from 'react-native';
import Svg, {Polyline} from 'react-native-svg';

interface Props {
  positive: boolean;
  children: number[];
}

interface State {
  width: number;
  height: number;
}

export default class SparkLine extends React.Component<Props, State> {
  state = {
    width: 0,
    height: 0
  }

  handleLayout: (event: LayoutChangeEvent) => void = ({nativeEvent: {layout: {width, height}}}) => {
    this.setState({
      width: width ? width : Dimensions.get('window').width,
      height
    });
  }

  render() {
    const {width, height} = this.state;
    const {positive, children}: Props = this.props;

    const maxValue = Math.max(...children);
    const minValue = Math.min(...children);
    const range = maxValue - minValue;
    const distance = width / (children.length - 1);
    const renderablePoints = children.map((v, i) => (
        {
        y: ( ( (v - minValue) * height ) / range ),
        x: distance * i + (i ? distance : 0)
        }
    )).map((v) =>  `${v.x},${-v.y + height}`);
    return (
        <View onLayout={this.handleLayout} style={styles.container}>
        {
          !width || !height ? null : (
              <Svg
                height={`${height}`}
                width={`${width}`}
              >
                <Polyline
                    points={renderablePoints.join(' ')}
                    fill="none"
                    stroke={positive ? positiveColor : negativeColor}
                    strokeWidth="1"
                />
              </Svg>
          )
        }
        </View>
    )
    }
}


const positiveColor = 'green';
const negativeColor = 'red';
const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'flex-end'
  },
  dot: {
    marginRight: 2,
    marginLeft: 2,
  },
});
