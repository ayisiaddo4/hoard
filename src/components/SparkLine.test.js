/* eslint-env jest */

import React from 'react';
import SparkLine from './SparkLine';

import renderer from 'react-test-renderer';

const MockedComponent = () => (
  <SparkLine positive={false} children={[1, 2, 3]} />
);

describe('<SparkLine />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<MockedComponent />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
