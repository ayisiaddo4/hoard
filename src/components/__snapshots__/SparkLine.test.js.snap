// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SparkLine /> renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "alignItems": "flex-end",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
/>
`;
